README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Draw Chernoff faces in ggplot2

[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/Selbosh/ggChernoff?branch=master&svg=true)](https://ci.appveyor.com/project/Selbosh/ggChernoff)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ggChernoff)](https://cran.r-project.org/package=ggChernoff)

This silly package, `ggChernoff`, introduces a `geom_chernoff` geom for
[`ggplot2`](http://ggplot2.org). This works a bit like
[`geom_point`](http://docs.ggplot2.org/current/geom_point.html), but
draws little smiley faces (emoticons) instead of dots.

The Chernoff geom has some unique aesthetics, including `smile`, which
makes your faces smile or frown according to the relative magnitude of
your continuous variable. By default, the mean value will generate a
straight face :expressionless: while higher values will make smiles
:blush: and lower values will draw frowns :cry:. You can customise this
using `scale_smile`. If `smile` is unmapped to a variable, all faces
will be happy by default.

Eyebrows are hidden by default, but you can activate them by mapping
something to the `brow` aesthetic. High values make your faces angry
:angry: and low values the opposite :anguished:.

## Getting started

Install the package using

``` r
devtools::install_github('Selbosh/ggChernoff')
```

and then load it using

``` r
library(ggChernoff)
```

## Examples

Firstly, let’s create a scatter plot of smiley faces out of Fisher’s
iris data set, each one coloured according to species.

``` r
library(ggplot2)
ggplot(iris) +
  aes(Petal.Width, Petal.Length, fill = Species) +
  geom_chernoff()
```

![](README-iris-1.png)<!-- -->

Here is an example using Immer’s barley data. We are happy about larger
yields\!

``` r
ggplot(lattice::barley) +
  aes(year, variety, smile = yield, brow = yield) +
  geom_chernoff(fill = 'goldenrod1') +
  scale_x_discrete(limits = c('1931', '1932')) +
  facet_wrap(~ site)
```

![](README-barley-1.png)<!-- -->

Basic legends are now supported. We can customise breaks and titles in
the usual `ggplot2` way, via `scale_smile_continuous`.

``` r
g <- ggplot(data.frame(x = rnorm(20), y = rexp(20), z = runif(20))) +
  aes(x, y, smile = z) +
  geom_chernoff(fill = 'steelblue1')
g
```

![](README-legends-1.png)<!-- -->

``` r
g + scale_smile_continuous('Smilez', breaks = 0:10/10, midpoint = .5)
```

![](README-legends-2.png)<!-- -->

You can also use this command to adjust the range of possible
happiness/sadness in your plot. In the following example, everybody is
somewhere between sad and straight-faced.

``` r
g + scale_smile_continuous(range = c(-1, 0))
```

![](README-range-1.png)<!-- -->

**New feature as of 0.3.0**: eye separation.

``` r
ggplot(iris) +
  aes(Petal.Width, Petal.Length, fill = Species, eyes = Sepal.Length) +
  geom_chernoff() +
  scale_eyes_continuous(range = c(0, 2))
```

![](README-eyes-1.png)<!-- -->

## Space invaders\! :alien:

``` r
cannon <- data.frame(x = 0, y = 0, colour = 'white', size = 20)
bunkers <- data.frame(x = seq(-4, 4, l = 4), y = 2, colour = 'green', size = 1)
ufos <- data.frame(x = rep(seq(-6, 6, length.out = 12), 5),
                   y = rep(6:10, each = 12), size = 10,
                   colour = c('cyan', 'yellow', 'magenta')[
                     c(rep(1:3,4), rep(c(2,3,1),4), rep(c(3,1,2),4), rep(1:3,4), rep(c(2,3,1),4))
                     ],
                   brow = rnorm(60))
ggplot(ufos) +
  aes(x, y, fill = colour, size = size) +
  geom_chernoff(smile = -1, aes(brow = brow), eyes = .5) +
  geom_chernoff(data = cannon) +
  geom_tile(data = bunkers, width = 1) +
  geom_tile(data = data.frame(x = 0, y = 3, colour = 'white', size = 2), width = .1) +
  scale_fill_identity() +
  scale_size_identity() +
  theme_void() +
  theme(plot.background = element_rect(fill = 'black'),
        legend.position = 'none')
#> Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
#> ℹ Please use `linewidth` instead.
```

![](README-spaceinvaders-1.png)<!-- -->

## References

Hermann Chernoff (1973). The use of faces to represent points in
*k*-dimensional space graphically. *Journal of the American Statistical
Association, 68*(342), 361–368.

Leland Wilkinson (2006). *The Grammar of Graphics* (2nd edition).
Springer.


NEWS.md
# ggChernoff 0.3.0

* Added an `eyes` aesthetic to control the distance between eyes.

# ggChernoff 0.2.0

* Initial submission to CRAN.

# ggChernoff 0.1.2.9001

* Added a `NEWS.md` file to track changes to the package.
* Tweaked `chernoffGrob` to make smiles more prominent, thanks to feedback from Lewis Rendell.

# ggChernoff 0.1.1.9001

* Added a more illustrative example for `scale_brow_continuous`.

# ggChernoff 0.1.1.9000

* Implemented data-driven eyebrows via the `brow` aesthetic.





R/geom_munch.R
#' Screaming faces in ggplot2
#'
#' @section Aesthetics:
#' \code{geom_chernoff} understands the following aesthetics (required aesthetics are in bold):
#' \itemize{
#' \item \strong{\code{x}}
#' \item \strong{\code{y}}
#' \item \code{colour}
#' \item \code{fill}
#' \item \code{size}
#' }
#' The following aesthetics are unique to \code{geom_chernoff}:
#' \itemize{
#' \item \code{scream}
#' }
#' For details, see \code{\link{screamGrob}}.
#'
#' @examples
#' library(ggplot2)
#' ggplot(iris, aes(Sepal.Width, Sepal.Length, scream = Petal.Length, fill = Species)) +
#'   geom_munch()
#'
#' @import ggplot2
#' @export geom_munch
geom_munch <- function(mapping = NULL, data = NULL, stat = "identity",
                       position = "identity", na.rm = FALSE, show.legend = NA,
                       inherit.aes = TRUE, ...) {
  layer(geom = GeomMunch, mapping = mapping, data = data, stat = stat,
        position = position, show.legend = show.legend, inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, ...)
        )
}

GeomMunch <- ggproto("GeomMunch", ggplot2::Geom,
  required_aes = c("x", "y"),
  default_aes = aes(colour = "black", fill = NA, size = 4, alpha = 1, scream = 1),
  draw_key = ggplot2::draw_key_rect,
  draw_panel = function(data, panel_scales, coord) {
    coords <- coord$transform(data, panel_scales)
    gt <- grobTree()
    for (i in seq_along(coords$x)) {
      gt <- addGrob(gt, screamGrob(coords$x[i],
                                   coords$y[i],
                                   coords$size[i],
                                   coords$colour[i],
                                   coords$fill[i],
                                   coords$alpha[i],
                                   coords$scream[i])
      )
    }
    return(gt)
  },
  draw_key = function(data, params, size) {
    screamGrob(x = .5, y = .5,
               data$size,
               data$colour,
               data$fill,
               data$alpha,
               data$scream)
  }
)


R/scale_brow.R
#' Scales for angry eyebrows
#'
#' \code{scale_brow} lets you customise how eyebrows are generated from your data.
#' It also lets you tweak the appearance of legends and so on.
#' By default, \code{brow} is set to \code{NA}, in which case no eyebrows will appear (see Examples).
#'
#' Use \code{range} to vary how angrily your maximum/minimum values are represented.
#' Minima smaller than -1 and maxima greater than +1 are possible but might look odd!
#' You can use \code{midpoint} to set a specific 'zero' value in your data or to have eyebrow angles represented as relative to average.
#'
#' The function \code{scale_brow} is an alias of \code{scale_brow_continuous}.
#' At some point we might also want to design a \code{scale_brow_discrete}, \code{scale_brow_manual} and so on.
#'
#' Legends are a work in progress. In particular, \code{size} mappings might produce odd results.
#'
#' @param ... Other arguments passed onto \code{\link[ggplot2]{continuous_scale}} to control name, limits, breaks, labels and so forth.
#' @param range Output range of eyebrow angles. +1 corresponds to very angry and -1 corresponds to a worried look.
#' @param midpoint A value or function of your data that will return level eyebrows, i.e. \code{¦:-)}
#'
#' @seealso \code{\link{geom_chernoff}}, \code{\link{scale_smile}}
#'
#' @importFrom scales rescale_mid
#'
#' @examples
#' library(ggplot2)
#' p <- ggplot(iris) +
#'     aes(Sepal.Width, Sepal.Length, fill = Species, brow = Sepal.Length) +
#'     geom_chernoff()
#' p
#' p + scale_brow_continuous(midpoint = min)
#' p + scale_brow_continuous(range = c(-.5, 2))
#'
#' # Only show eyebrows if 'sad', otherwise hide them
#' usa <- data.frame(date = c(time(presidents)), rating = c(presidents))
#' ggplot(subset(usa, complete.cases(usa))) +
#'     aes(date, rating, smile = rating, fill = rating,
#'         brow = ifelse(rating < 50, rating, NA)) +
#'     geom_line() +
#'     geom_chernoff(show.legend = FALSE) +
#'     scale_brow(range = -1:0) +
#'     scale_fill_gradient(low = 'skyblue1', high = 'goldenrod1')
#'
#'
#' @rdname scale_brow
#' 
#' @return
#' A \code{\link[ggplot2:continuous_scale]{Scale}} layer object for use with \code{ggplot2}.
#'
#' @export
scale_brow_continuous <- function(..., range = c(-1, 1), midpoint = mean) {
  if (is.numeric(midpoint)) {
    neutral <- function(...) return(midpoint)
  } else {
    neutral <- match.fun(midpoint)
  }
  continuous_scale('brow', 'brow_c',
                   function(x) scales::rescale_mid(x, to = range, mid = neutral(x, na.rm = TRUE)),
                   ..., na.value = NA)
}

#' @rdname scale_brow
#' @export
scale_brow <- scale_brow_continuous


R/scale_eyes.R
#' Scales for eye separation
#'
#' \code{scale_eyes} lets you customise how eye separation is determined from your data.
#' It also lets you tweak the appearance of legends and so on.
#'
#' Use \code{range} to vary how happily/sadly your maximum/minimum values are represented.
#' Minima smaller than -1 and maxima greater than +1 are possible but might look odd!
#' You can use \code{midpoint} to set a specific 'zero' value in your data or to have eye width represented as relative to average.
#'
#' The function \code{scale_eyes} is an alias of \code{scale_eyes_continuous}.
#'
#' Legends are a work in progress. In particular, \code{size} mappings might produce odd results.
#'
#' @param ... Other arguments passed onto \code{\link[ggplot2]{continuous_scale}} to control name, limits, breaks, labels and so forth.
#' @param range Output range of eye distances. 0 corresponds to a cyclops and +1 to a 'normal' distance.
#' @param midpoint A value or function of your data that will return a 'normal' separation
#'
#' @seealso \code{\link{geom_chernoff}}, \code{\link{scale_brow}}, \code{\link{scale_smile}}
#'
#' @importFrom scales rescale_mid
#'
#' @examples
#' library(ggplot2)
#' p <- ggplot(iris) +
#'     aes(Sepal.Width, Sepal.Length, fill = Species, eyes = Sepal.Length) +
#'     geom_chernoff()
#' p
#' p + scale_eyes_continuous(midpoint = min)
#' p + scale_eyes_continuous(range = c(0, 2))
#'
#' @rdname scale_eyes
#' 
#' @return
#' A \code{\link[ggplot2:continuous_scale]{Scale}} layer object for use with \code{ggplot2}.
#'
#' @export
scale_eyes_continuous <- function(..., range = c(.1, 2), midpoint = mean) {
  if (is.numeric(midpoint)) {
    neutral <- function(...) return(midpoint)
  } else {
    neutral <- match.fun(midpoint)
  }
  continuous_scale('eyes', 'eyes_c',
                   function(x) scales::rescale_mid(x, to = range, mid = neutral(x, na.rm = TRUE)),
                   ..., na.value = 1)
}

#' @rdname scale_eyes
#' @export
scale_eyes <- scale_eyes_continuous


R/screamGrob.R
#' Draw a screaming face
#'
#' Generates a face in the style of Edvard Munch's 1893 painting \emph{The Scream}.
#' Adjusting parameters of the function can make the expression more or less dramatic.
#'
#' @param x horizontal position
#' @param y vertical position.
#' @param size area of the face
#' @param colour colour of outlines and features
#' @param alpha transparency, where 0 is transparent and 1 is opaque
#' @param scream the higher this value, the more dramatic the expression
#'
#' @return A \code{\link[grid]{grobTree}} object.
#'
#' @references
#' Murrell, Paul (2012). It's Not What You Draw, It's What You Don't Draw. \emph{R Journal} 4/2. \url{https://journal.r-project.org/archive/2012-2/RJournal_2012-2_Murrell2.pdf}
#'
#' Munch, Edvard (1893). \emph{Skrik (Der Schrei der Natur)}.
#'
#' @examples
#' grid::grid.newpage()
#' grid::grid.draw(screamGrob(.7, .7, size = 50))
#' grid::grid.draw(screamGrob(size = 50, fill = 'goldenrod1'))
#'
#' @seealso \code{\link{chernoffGrob}}
#' @import grid
#' @importFrom gridExtra ellipseGrob
#' @export
screamGrob <- function(x = .5, y = .5,
                       size = 4,
                       colour = 'black',
                       fill = NA,
                       alpha = 1,
                       scream = 1) {
  .pt <- 72.27 / 25.4
  vp <- viewport(x = x, y = y,
                 width = unit(sqrt(size * .pt), 'mm'),
                 height = unit(sqrt(size * .pt), 'mm'),
                 default.units = 'npc')
  facex <- c(.1, .1, .9, .9, # Cranium
             .9, .9, .7, .7, # Right cheek
             .7, .7, .3, .3, # Chin
             .3, .3, .1, .1) # Left cheek
  facey <- c(.7,  1,  1, .7, # Cranium
             .7, .5, .6, .3, # Right cheek
             .3, .1, .1, .3, # Chin
             .3, .6, .5, .7) # Left cheek
  faceShape <- bezierGrob(facex, facey,
                          id.lengths = rep(4, 4),
                          default.units = 'npc',
                          vp = vp)
  facePoints <- do.call(rbind, bezierPoints(faceShape))
  faceGrob <- polygonGrob(convertUnit(unit(unlist(facePoints[, 1]), 'inches'), 'npc'),
                          convertUnit(unit(unlist(facePoints[, 2]), 'inches'), 'npc'),
                          gp = gpar(colour = colour, fill = fill),
                          vp = vp)
  eyesGrob <- circleGrob(c(.3, .7),
                         c(.7, .7) * 2,
                         r = .1,
                         default.units = 'npc',
                         gp = gpar(colour = colour, fill = colour),
                         vp = vp)
  mouthGrob <- gridExtra::ellipseGrob(.5,
                                      .35 * 2,
                                      angle = 0,
                                      size = .1,
                                      position.units = 'npc',
                                      size.units = 'npc',
                                      ar = .6,
                                      gp = gpar(colour = colour, fill = colour),
                                      vp = vp)
  grobTree(faceGrob, eyesGrob, mouthGrob)
}


R/scale_smile.R
#' Scales for smiling and frowning
#'
#' \code{scale_smile} lets you customise how smiles are generated from your data.
#' It also lets you tweak the appearance of legends and so on.
#'
#' Use \code{range} to vary how happily/sadly your maximum/minimum values are represented.
#' Minima smaller than -1 and maxima greater than +1 are possible but might look odd!
#' You can use \code{midpoint} to set a specific 'zero' value in your data or to have smiles represented as relative to average.
#'
#' The function \code{scale_smile} is an alias of \code{scale_smile_continuous}.
#' At some point we might also want to design a \code{scale_smile_discrete}, \code{scale_smile_manual} and so on.
#'
#' Legends are a work in progress. In particular, \code{size} mappings might produce odd results.
#'
#' @param ... Other arguments passed onto \code{\link[ggplot2]{continuous_scale}} to control name, limits, breaks, labels and so forth.
#' @param range Output range of smiles. +1 corresponds to a full smile and -1 corresponds to a full frown.
#' @param midpoint A value or function of your data that will return a neutral/straight face, i.e. \code{:-|}
#'
#' @seealso \code{\link{geom_chernoff}}, \code{\link{scale_brow}}
#'
#' @importFrom scales rescale_mid
#'
#' @examples
#' library(ggplot2)
#' p <- ggplot(iris) +
#'     aes(Sepal.Width, Sepal.Length, fill = Species, smile = Sepal.Length) +
#'     geom_chernoff()
#' p
#' p + scale_smile_continuous(midpoint = min)
#' p + scale_smile_continuous(range = c(-.5, 2))
#'
#' @rdname scale_smile
#' 
#' @return
#' A \code{\link[ggplot2:ggplot2-ggproto]{Scale}} layer object for use with \code{ggplot2}.
#'
#' @export
scale_smile_continuous <- function(..., range = c(-1, 1), midpoint = mean) {
  if (is.numeric(midpoint)) {
    neutral <- function(...) return(midpoint)
  } else {
    neutral <- match.fun(midpoint)
  }
  continuous_scale('smile', 'smile_c',
                   function(x) scales::rescale_mid(x, to = range, mid = neutral(x, na.rm = TRUE)),
                   ..., na.value = 1)
}

#' @rdname scale_smile
#' @export
scale_smile <- scale_smile_continuous


R/chernoffGrob.R
#' Draw a smiley face
#'
#' Uses \code{\link[grid]{Grid}} graphics to draw a face.
#'
#' @param x horizontal position
#' @param y vertical position
#' @param size area of the face
#' @param colour colour of outlines and features
#' @param fill fill colour
#' @param alpha transparency, where 0 is transparent and 1 is opaque
#' @param smile amount of smiling/frowning
#' @param brow eyebrow angle, to represent anger or concern
#' @param nose logical. Adds a nose to the face
#' @param eyes distance between the eyes
#'
#' @return A \code{\link[grid]{grobTree}} object.
#'
#' @seealso \code{\link{geom_chernoff}}
#'
#' @import grid
#' @export
#'
#' @examples
#' face <- chernoffGrob(.5, .5, size = 1e3, smile = -1, brow = 1, colour = 'navy', fill = 'lightblue')
#' grid::grid.newpage()
#' grid::grid.draw(face)
chernoffGrob <- function(x = .5, y = .5,
                         size = 1,
                         colour = 'black',
                         fill = NA,
                         alpha = 1,
                         smile = 1,
                         brow = NA,
                         nose = FALSE,
                         eyes = 1) {
  .pt <- 72.27 / 25.4
  faceGrob <- circleGrob(x, y, r = unit(sqrt(.5 * size * .pt), 'mm'))
  vp1 <- viewport(x = x, y = y,
                  width = grobWidth(faceGrob), height = grobHeight(faceGrob))
  eyesGrob <- circleGrob(rep(0.5 + eyes * c(-.2, +.2), each = length(x)),
                         .6,
                         r = 1/20,
                         gp = gpar(fill = colour),
                         vp = vp1)
  if (!is.na(brow)) {
    browGrob <- polylineGrob(x = c(.2, .4, .6, .8),
                             y = .75 + brow * c(+.05, -.05, -.05, +.05), # .7--.8
                             id = rep(1:2, each = 2),
                             gp = gpar(col = colour),
                             vp = vp1)
  } else browGrob <- nullGrob()
  noseGrob <- circleGrob(r = 1/15,
                         gp = gpar(col = ifelse(nose, colour, NA), fill = NA),
                         vp = vp1)
  mouthGrob <- bezierGrob(rep(.5, each = 4) + c(-.25, -.1, .1, .25),
                          rep(.28, each = 4) + smile * c(.08, -.12, -.12, .08),
                          gp = gpar(fill = colour),
                          id.lengths = rep(4, length(x)),
                          vp = vp1)
  grobTree(faceGrob, noseGrob, eyesGrob, browGrob, mouthGrob,
           gp = gpar(alpha = alpha, col = colour, fill = fill)
           )
}


R/geom_chernoff.R
#' Chernoff faces in ggplot2
#'
#' The Chernoff geom is used to create data visualisations in the shape of human-like faces.
#' By mapping to the relevant aesthetics, faces can appear to vary in happiness, anger, size, colour and so on.
#'
#' @param mapping Set of aesthetic mappings created by \code{\link[ggplot2]{aes}} or
#'   \code{\link[ggplot2]{aes_}}. If specified and \code{inherit.aes = TRUE} (the
#'   default), is combined with the default mapping at the top level of the
#'   plot. You only need to supply \code{mapping} if there isn't a mapping
#'   defined for the plot.
#' @inheritParams ggplot2::geom_point
#'
#' @section Aesthetics:
#' \code{geom_chernoff} understands the following aesthetics (required aesthetics are in bold):
#' \itemize{
#' \item \strong{\code{x}}
#' \item \strong{\code{y}}
#' \item \code{colour}
#' \item \code{fill}
#' \item \code{size}
#' }
#' The following aesthetics are unique to \code{geom_chernoff}:
#' \itemize{
#' \item \code{smile}
#' \item \code{brow}
#' \item \code{nose}
#' \item \code{eyes}
#' }
#' For details, see \code{\link{chernoffGrob}}.
#'
#' @examples
#' library(ggplot2)
#' ggplot(iris, aes(Sepal.Width, Sepal.Length, smile = Petal.Length, fill = Species)) +
#'   geom_chernoff()
#'
#' ggplot(data.frame(x = 1:4,
#'                   y = c(3:1, 2.5),
#'                   z = factor(1:4),
#'                   w = rnorm(4),
#'                   n = c(rep(FALSE, 3), TRUE)
#'                   )) +
#'     aes(x, y, fill = z, size = x, nose = n, smile = w) +
#'     geom_chernoff()
#'
#' @references
#' Chernoff, H. (1973).
#' The use of faces to represent points in \emph{k}-dimensional space graphically.
#' \emph{Journal of the American Statistical Association, 68}(342), 361–368.
#'
#' @seealso \code{\link{chernoffGrob}}
#' 
#' @return
#' A \code{\link[ggplot2:Geom]{Geom}} layer object for use with \code{ggplot2}.
#'
#' @import ggplot2
#' @export geom_chernoff
geom_chernoff <- function(mapping = NULL, data = NULL, stat = "identity",
                          position = "identity", na.rm = FALSE, show.legend = NA,
                          inherit.aes = TRUE, ...) {
  layer(geom = GeomChernoff, mapping = mapping, data = data, stat = stat,
        position = position, show.legend = show.legend, inherit.aes = inherit.aes,
        params = list(na.rm = na.rm, ...)
        )
}

GeomChernoff <- ggproto("GeomChernoff", ggplot2::Geom,
  required_aes = c("x", "y"),
  default_aes = aes(colour = "black", fill = NA, size = 4, alpha = 1, smile = 1, brow = NA, nose = FALSE,
                    eyes = 1),
  draw_key = ggplot2::draw_key_rect,
  draw_panel = function(data, panel_scales, coord) {
      coords <- coord$transform(data, panel_scales)
      gl <- grobTree()
      for (i in seq_along(coords$x)) {
        # Filthy hack: draw one whole face at a time
        # so overlapping faces are rendered correctly.
        gl <- addGrob(gl, chernoffGrob(coords$x[i],
                                       coords$y[i],
                                       coords$size[i],
                                       coords$colour[i],
                                       coords$fill[i],
                                       coords$alpha[i],
                                       coords$smile[i],
                                       coords$brow[i],
                                       coords$nose[i],
                                       coords$eyes[i])
                      )
      }
      return(gl)
  },
  draw_key = function(data, params, size) {
    chernoffGrob(x = .5, y = .5,
                 data$size,
                 data$colour,
                 data$fill,
                 data$alpha,
                 data$smile,
                 data$brow,
                 data$nose,
                 data$eyes)
  }
)
